packages/instrumentation-amqplib/src/amqplib.ts

```diff
@@ -0,0 +1,377 @@
+                    onMessage.call(this, msg);
+                });
+
+                if (options?.noAck) {
+                    self.callConsumeEndHook(span, msg, false, EndOperation.AutoAck);
+                    span.end();
+                }
+            };
+            arguments[1] = patchedOnMessage;
+            return original.apply(this, arguments);
+        };
+    }
+
+    private _getPublishPatch(
+        moduleVersion: string,
+        original: (exchange: string, routingKey: string, content: Buffer, options?: amqp.Options.Publish) => boolean
+    ) {
+        const self = this;
+        return function publish(
+            exchange: string,
+            routingKey: string,
+            content: Buffer,
+            options?: amqp.Options.Publish
+        ): boolean {
+            const normalizedExchange = normalizeExchange(exchange);
+            const span = self.tracer.startSpan(`${normalizedExchange} -> ${routingKey} send`, {
+                kind: SpanKind.PRODUCER,
+                attributes: {
+                    ...this.connection[CONNECTION_ATTRIBUTES],
+                    [MessagingAttribute.MESSAGING_DESTINATION]: exchange,
+                    [MessagingAttribute.MESSAGING_DESTINATION_KIND]: 'topic',
+                    [MessagingAttribute.MESSAGING_RABBITMQ_ROUTING_KEY]: routingKey,
+                },
+            });
+
+            if (self._config.moduleVersionAttributeName) {
+                span.setAttribute(self._config.moduleVersionAttributeName, moduleVersion);
+            }
+
+            const modifiedOptions = options ?? {};
+            modifiedOptions.headers = modifiedOptions.headers ?? {};
+            propagation.inject(setSpan(context.active(), span), modifiedOptions.headers);
+
+            if (self._config.publishHook) {
+                safeExecuteInTheMiddle(
+                    () => self._config.publishHook(span, { exchange, routingKey, content, options }),
+                    (e) => {
+                        if (e) {
+                            diag.error('amqplib instrumentation: publishHook error', e);
+                        }
+                    },
+                    true
+                );
+            }
+
+            // calling original publish function is only storing the message in queue.
+            // it does not send it and waits for an ack, so the span duration is expected to be very short.
+            const argumentsCopy = [...arguments];
+            argumentsCopy[3] = modifiedOptions;
+            const originalRes = original.apply(this, argumentsCopy);
+            if (!originalRes) {
+                span.setStatus({
+                    code: SpanStatusCode.ERROR,
+                    message: 'write buffer is full, message should be published again after drain event',
+                });
+            }
+            span.end();
+            return originalRes;
+        };
+    }
+
+    private endConsumerSpan(
```

Comment:
some private funcs name start with `_` and some don't, better choose one and stick with it